pandas/tests/io/test_html.py

```diff
@@ -660,14 +940,15 @@ def test_parse_dates_combine(self):
 
     def test_computer_sales_page(self):
         data = os.path.join(DATA_PATH, 'computer_sales_page.html')
-        with tm.assert_raises_regex(ParserError,
-                                    r"Passed header=\[0,1\] are "
-                                    r"too many rows for this "
-                                    r"multi_index of columns"):
-            self.read_html(data, header=[0, 1])
+        # This table is unique because it has no <thead>, and its <th>-only
```

Comment:
Hmm not sure about this. Certainly a crazy example but this changes the behavior of the arguments passed to `header` now, no?

Reply:
This pull request doesn't change the behavior of the `header` argument. It changes the behavior one gets when the `header` argument is _omitted_. That change has no effect on this specific test.

When it comes to this specific unit test ("Test some old table in the wild"):

* Prior to this pull request, pandas erroneously crashed on this table. The unit test was bad -- it tested for a bug.
* This pull request includes a hack involving th-only rows. That hack would _seem_ to infer the correct `header=[0,1]` for this specific table, but it doesn't because this specific table is weird

This unit test now tests that pandas _works_, instead of testing that it is broken. The comment explains why the `header=[0,1]` is needed -- since the HTML makes it look like it shouldn't be.

Reply:
It looks like it broke previously for `header=[0, 1]` but worked for `header=[1, 2]`. Now it is working for `header=[0,1]` and that returns the same value that previously was returned with `header=[1,2]`, right?

Not saying that is necessarily a bad thing but looking for confirmation as in the off-chance something like this does occur in the wild that is a pretty subtle change for users that may have been expecting the former behavior

Reply:
No, that's not what changed. It broke before because it didn't support rowspan/colspan. The deletion of empty rows is existing behavior.

Reply:
urgh, looking at the diff with master instead of my own. I see that `header=[1,2]` used to not crash. I'll investigate more to figure out exactly why that worked.

Reply:
Okay, you're absolutely right -- the behavior changed.

Previously, pandas removed all-empty rows from the `<thead>` but _not_ from the `<tbody>`.

Now, pandas removes all-empty rows from anywhere in the table.

So `header` changed on this table because its header rows are inside the `<tbody>`.

Should I revert that change? It was unintentional. But the original behavior appears to be unintentional to me, too -- why would `<thead>` and `<tbody>` behave differently?

Reply:
References:

* #13434 suggests read_html() should infer multiple table rows
* #15242 implements by deleting all-empty rows from `<thead>`

From these discussions, I gather that pandas shouldn't auto-delete data, but auto-deleting in the `<thead>` is kinda okay because otherwise we'd get a ParserError by default.

I'll auto-delete empty header rows. It seems to me `_data_to_frame()` is the correct place to do that. @WillAyd does that sound appropriate? Then I can revert the change to this unit test -- I was misunderstanding the error.

Reply:
Fixed in commit ed120d8. I added new tests for behavior implied in #15242.

Unfortunately, `test_spam_header` tested for a bug -- it specified `header=1` instead of the intent, `header=2`. The test passed because of a bug in the previous implementation: indexing happened _after_ deleting empty header rows.

Let's track the bugs through history to see why `header=2` is correct:

1. Original implementation, commit b31c033f (2013) -- uses `header=0` to select `<td>` elements within `<thead>` that are not part of a `<tr>`. The bug that made this work: pandas erroneously treated `<thead>` as equivalent to `<tr>`, merging all `<th>` and `<td>` within the `<thead>` into one big row.
2. commit a8723a4b (2013): changed the test to `header=1` and changed the value it's looking for to be the value from the first `<tr>` in the `<tbody>`. Seems like a silly change. The author probably tested a body cell because lxml and bs4 began treating the (malformed) header differently as a result of some completely-unrelated change.
3. commit 0ab081345 (#15242, 2017): kept `header=1` unchanged, because two new behaviors cancelled out. First, `<thead>` handling was corrected to look for `<tr>`s inside the `<thead>` (meaning one would need to set `header=2`). Second, `_parse_raw_thead` was _incorrectly_ changed to delete empty rows (deleting one row from the header brings us back to `header=1`). That change made the unit test pass, but the unit test was testing the wrong behavior all along.
4. Today, I fixed the bug introduced in 2017, meaning `header=2` finds the cell that's been referenced since a8723a4b in 2013.

@WillAyd what do you think of this change? I think anybody would agree the previous behavior was a bug. And it seems silly to re-implement the bug in this pull request in the name of avoiding feature creep.

I imagine this applies to only a tiny number of tables in the wild.

Reply:
Thanks for the clarification. I do agree that `header=2` appears to be more correct. Do you think there's an issue with other index-based keywords (like `skiprows`) in the current implementation? 

OK to address in a separate issue but it would be nice to have mention of this in a whatsnew in some fashion (again separate issue is fine)